internal/services/applications/application_data_source_msgraph.go

```diff
@@ -0,0 +1,127 @@
+			fieldValue = applicationId
+		} else if displayName, ok := d.Get("display_name").(string); ok && displayName != "" {
+			fieldName = "displayName"
+			fieldValue = displayName
+		} else if name, ok := d.Get("name").(string); ok && name != "" {
+			fieldName = "displayName"
+			fieldValue = name
+		} else {
+			return tf.ErrorDiagF(nil, "One of `object_id`, `application_id` or `displayName` must be specified")
+		}
+
+		filter := fmt.Sprintf("%s eq '%s'", fieldName, fieldValue)
+
+		result, _, err := client.List(ctx, filter)
+		if err != nil {
+			return tf.ErrorDiagF(err, "Listing applications for filter %q", filter)
+		}
+
+		switch {
+		case result == nil || len(*result) == 0:
+			return tf.ErrorDiagF(fmt.Errorf("No applications found matching filter: %q", filter), "Application not found")
+		case len(*result) > 1:
+			return tf.ErrorDiagF(fmt.Errorf("Found multiple applications matching filter: %q", filter), "Multiple applications found")
+		}
+
+		app = &(*result)[0]
+		switch fieldName {
+		case "appId":
+			if app.AppId == nil {
+				return tf.ErrorDiagF(fmt.Errorf("nil AppID for applications matching filter: %q", filter), "Bad API Response")
+			}
+			if *app.AppId != fieldValue {
+				return tf.ErrorDiagF(fmt.Errorf("AppID does not match (%q != %q) for applications matching filter: %q", *app.AppId, fieldValue, filter), "Bad API Response")
+			}
+		case "displayName":
+			if app.DisplayName == nil {
+				return tf.ErrorDiagF(fmt.Errorf("nil displayName for applications matching filter: %q", filter), "Bad API Response")
+			}
+			if *app.DisplayName != fieldValue {
+				return tf.ErrorDiagF(fmt.Errorf("DisplayName does not match (%q != %q) for applications matching filter: %q", *app.DisplayName, fieldValue, filter), "Bad API Response")
+			}
+		}
+	}
+
+	if app.ID == nil {
+		return tf.ErrorDiagF(fmt.Errorf("Object ID returned for application is nil"), "Bad API Response")
+	}
+
+	d.SetId(*app.ID)
+
+	tf.Set(d, "app_roles", helpers.ApplicationFlattenAppRoles(app.AppRoles))
+	tf.Set(d, "application_id", app.AppId)
+	tf.Set(d, "available_to_other_tenants", app.SignInAudience == msgraph.SignInAudienceAzureADMultipleOrgs)
+	tf.Set(d, "display_name", app.DisplayName)
+	tf.Set(d, "group_membership_claims", app.GroupMembershipClaims)
+	tf.Set(d, "identifier_uris", tf.FlattenStringSlicePtr(app.IdentifierUris))
+	tf.Set(d, "name", app.DisplayName) // TODO: remove in v2.0
+	tf.Set(d, "object_id", app.ID)
+	tf.Set(d, "optional_claims", flattenApplicationOptionalClaims(app.OptionalClaims))
+	tf.Set(d, "required_resource_access", flattenApplicationRequiredResourceAccess(app.RequiredResourceAccess))
+
+	// TODO: remove `type` in v2.0
+	var appType string
+	if v := app.IsFallbackPublicClient; v != nil && *v {
+		appType = "native"
+	} else {
+		appType = "webapp/api"
+	}
+	tf.Set(d, "type", appType)
+
+	if app.Api != nil {
+		tf.Set(d, "oauth2_permissions", helpers.ApplicationFlattenOAuth2Permissions(app.Api.OAuth2PermissionScopes))
+	}
```

Comment:
I've opted to work around this in the Read func for now, since the schema in the next version will mirror this with nested blocks and the flatten funcs will change then.